### ISSUE TITLE
metis doesn't always respect cflags

### ISSUE THREAD

**Issue by KineticTheory** [spack/spack#6839]
### Summary

The current recipe for metis allows conflicting optimization levels to appear on the compile line in a way that isn't expected. For example: `spack -d install metis build_type=RelWithDebInfo cflags="-O3 -g"` doesn't compile with the expected CFLAGS.  I expect to see compile lines similar to this:

```
/usr/bin/gcc -O3 -g <include directives and defines> -o file.o -c file.c
```

### Actual Result

```
/usr/bin/gcc -O3 -g <include directives and defines> -O3 -O2 -g -DDNDEBUG -o file.o -c file.c
```

Because `-O2` is listed after `-O3` the optimization is not what I expect.  I believe the `-O2 -g -DDNDEBUG` flags are inserted by CMake as they are the default `CMAKE_C__FLAGS_RELWITHDEBINFO` values for gcc.

I'm actually not sure where that second `-O3` is coming from.

### Possible solution

If `SPACK_CFLAGS` is set in the build environment, then zero-out the cmake defaulted `CMAKE_C_FLAGS_XXX` variables.

I have a patch incoming (#6840) to fix this problem (if it passes review).

### Related Issues

#6732 
#6640 
#6749 
#6745 
#6714

**Comment by scheibelp** [comment]
The solution suggested at https://github.com/spack/spack/pull/6714#issuecomment-352630032 may be appropriate here (i.e. use the new flag handler logic to tell the build system what flags to use).

**Comment by KineticTheory** [comment]
@scheibelp I tried to copy some of the `flag_handler` options described in other tickets but I couldn't make anything work the way I wanted.  I'm not particularly well versed in python or the main spack code, so I might have implemented this approach incorrectly.  Can you suggest actual code changes to `metis/package.py` that implement your suggestion?

**Comment by scheibelp** [comment]
Is there an issue with converting to `CMakePackage` or with setting the `flag_handler`? One approach which *may* work is to just set `flag_handler = env_flags` (normally for CMake packages I would recommend `flag_handler = build_system_flags`); in that case you would not have to update the `metis` package definition to be an instance of `CMakePackage` and in fact that might be the only change necessary.

**Comment by KineticTheory** [comment]
I don't think we can make metis a CMakePackage since Metis prior to version 5 does not use cmake and a lot of folks still use metis@4.0.3.

I will try the `flag_hander` options that you suggest above and report the results here.

**Comment by KineticTheory** [comment]
In the metis recipe, I added the `flag_handler` like this (is this correct?):

```diff
    @when('@5:')
    def install(self, spec, prefix):
        source_directory = self.stage.source_path
        build_directory = join_path(source_directory, 'build')

        options = std_cmake_args[:]
+       self.flag_handler = self.build_system_flags
```
Now, install metis: `spack -d install metis cflags="-O0 -g"` (note that `build_type` defaults to _Release_).
However, the actual compile commands are similar to this: `/usr/bin/gcc <rpaths> -O0 -g <defines> -O3 -O3 -DNDEBUG -o file.o -c file.c`

This is not the behavior that I expect (an optimized metis is created instead of `-O0`).  If `cflags` are explicitly set, then default cmake compile flags should be ignored (even when `CMAKE_BUILD_TYPE=Release`) .

**Comment by scheibelp** [comment]
Sorry @KineticTheory to be clearer: you would have to set `flag_handler` at the class definition level rather than in the `install` method. Something like:

```
    flag_handler = build_system_flags

    @when('@5:')
    def install(self, spec, prefix):
        source_directory = self.stage.source_path
        build_directory = join_path(source_directory, 'build')

        options = std_cmake_args[:]
```

Note also that you can only set `flag_handler = build_system_flags` (note you don't need to prefix with `self`) if you have updated `metis` to be an instance of `CMakePackage`. Otherwise I would suggest setting `flag_handler = env_flags` (which might not work but from what I read CMake can use environment variables like `CFLAGS`).

Let me know if that doesn't work out.

**Comment by KineticTheory** [comment]
@scheibelp I moved the `flag_handler` up to class scope and out of the `install` function:

```diff
@@ -75,6 +75,8 @@ class Metis(Package):
         url += "/metis-{0}.tar.gz".format(version)
         return url
 
+    flag_handler = env_flags
+
     @when('@5:')
     def patch(self):
         source_path = self.stage.source_path
```
 
but the results are the same as above.  

Related to #6860, I also tried adding `flag_hander =  build_system_flags` to `parmetis/package.py` (a CMakePackage) but spack didn't like that command and resulted in the error: `NameError: name 'build_system_flags' is not defined`.  This function is defined at `lib/spack/spack/package.py` but for some reason the recipe can't see it.

Are there any TPL recipes that currently use these `flag_hander` functions?  I would like to see an example.

**Comment by scheibelp** [comment]
> Are there any TPL recipes that currently use these flag_hander functions? I would like to see an example.

There are not actually any packages using this at the moment. The feature is new. The documentation describes the feature in more detail at: http://spack.readthedocs.io/en/latest/packaging_guide.html#compiler-flags

> I also tried adding flag_hander = build_system_flags to parmetis/package.py (a CMakePackage) but spack didn't like that command and resulted in the error: `NameError: name 'build_system_flags' is not defined`. This function is defined at `lib/spack/spack/package.py` but for some reason the recipe can't see it.

I was expecting it wasn't required but it looks like you will have to set

`flag_handler = Package.build_system_flags`

(which I should have checked in the linked documentation). 

> the results are the same as above [when using env_flags for metis]

I can confirm that; the compiler invocations made by the build system include both `-O0` and `-O3`, so this isn't a matter of a conflict between compiler flags added by the build system and injected by Spack. Unfortunately it looks like the build system isn't doing any work to check whether there is overlap between the `CFLAGS` environment variable and flags that it chooses.

I'd have to think on how this might be handled here. For `parmetis` things should be more straightforward since it inherits from `CMakePackage`

Adding @becker33: FYI in this case the package cannot inherit from `CMakePackage` because not all versions support cmake, but it appears that for versions that do support cmake they could use `CMakePackage`'s functionality (including `flags_to_build_system_args`). I don't assume this has a simple fix but IMO it is an interesting use case.

**Comment by KineticTheory** [comment]
This issue will need to be reviewed and either updated or closed after #6960 and #6961 are closed/merged.
